c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -482,19 +484,131 @@ class CachingAllocatorConfig {
+  bool active = false; // if there is plan for at least one device
+  void create_plan(
+      const std::vector<std::vector<AllocFreeEvent>>& plan,
+      std::vector<Block*> plan_ptrs,
+      std::vector<size_t> plan_size,
+      std::vector<uint32_t> non_rec_lengths) {
+    std::lock_guard<std::mutex> lock(mutex);
+    mem_plan = plan;
+    plan_blk_ptr.resize(plan.size(), nullptr);
+    plan_counter.resize(plan.size(), 1);
+    planning_size.resize(plan.size(), 0);
+    allocated_blocks.resize(plan.size());
+    non_rec_length.resize(plan.size(), 0);
+    plan_blk_ptr = plan_ptrs;
+    planning_size = plan_size;
+    non_rec_length = non_rec_lengths;
+    active = true;
+  }
+
+  Block* malloc(size_t size, int device, cudaStream_t stream) {
+    Block* block =
+        nullptr; // if block remains null, this indicates event not allocated by
+                 // plane, hence, use CudaCachingAllocator
+    // event is allocated according to plan if its size matches the expected
+    // allocation size
+    if ((plan_blk_ptr[device] != nullptr) &&
+        (size == (uint32_t)mem_plan[device][plan_counter[device]].size)) {
+      intptr_t planned_ptr = mem_plan[device][plan_counter[device]]
+                                 .ptr; // allocation offset according to plan
+      auto it_ptr_lower_bound = allocated_blocks[device].lower_bound(
+          planned_ptr); // find the closest start/end point of a previous
+                        // allocation after the current allocate ptr
+      // perform the allocation only if it fails inside the block assigned for
+      // plan execution and it does not overlap with previuosly allocated events
+      // if it_ptr_lower_bound points to an end point, this means the current
+      // allocation starts in the middle of a previous allocation; do not
+      // allocate. If it_ptr_lower_bound points to a start point, check if
+      // planned_ptr + size <= start point to ensure no overlap
+      if ((planned_ptr + size <= planning_size[device]) &&
+          ((it_ptr_lower_bound == allocated_blocks[device].end()) ||
+           (it_ptr_lower_bound->second &&
+            (planned_ptr + size) <= (uint32_t)it_ptr_lower_bound->first))) {
+        block = new Block(
+            device,
+            stream,
+            size,
+            nullptr,
+            reinterpret_cast<void*>(
+                reinterpret_cast<intptr_t>(plan_blk_ptr[device]->ptr) +
+                planned_ptr)); // allocation block ptr is planning block ptr +
+                               // allocation offset according to plan
+        block->planned = true; // marks the block as allocated by plan
+        allocated_blocks[device][planned_ptr] =
+            true; // store start point of allocation
+        allocated_blocks[device][planned_ptr + size - 1] =
+            false; // store end point of allocation
+      }
+      plan_counter[device] += 1; // wait for next allocation
```

Comment:
this means we move forward regardless of the current allocation is expected or not?

Reply:
Since this is inside "if ((plan_blk_ptr[device] != nullptr) && (size == (uint32_t)mem_plan[device][plan_counter[device]].size))", we only move if the allocation is expected? However, if the allocation is expected but not allocated according to plan, we move forward. The later case can happen in two scenarios:
1. Plan is not feasible.
2.  Unexpected event with size and order that matches the expected event is allocated according to plan but freed later than expected by plan.